[docs] Publish the Jenkins Plugin blog post (#5329)
DevExpress/testcafe: docs/articles/blog/2020-7-15-introducing-the-testcafe-jenkins-plugin.md

---
layout: post

> title: "Introducing the TestCafe Jenkins plugin"

Reviewer: ```suggestion
title: "Introducing the TestCafe Jenkins Plugin"
```

permalink: /media/team-blog/:title.html
isTeamBlog: true

> author: Sergey Shurygin, Eugene Titerman

Reviewer 2: Please change first author's name to
```Pavel Redrukhin```

---

> # Introducing the TestCafe Jenkins plugin

Reviewer: ```suggestion
# Introducing the TestCafe Jenkins Plugin
```

> TestCafe can capture videos and screenshots so you can debug your tests and examine page UI. You can record all the tests or only those that failed, generate multiple video files or stitch them together, and even single out specific interactions (see [the documentation](https://devexpress.github.io/testcafe/documentation/guides/advanced-guides/screenshots-and-videos.html) for the full list of available options).

Reviewer: What do you mean by "stitch them together"? Is a single video file generated for multiple tests? If so, can you phrase it this way instead?

Author: Would `You can record all the tests or only those that failed, generate unique video files for each test or record all of them in one take, and even single out specific interactions (see [the documentation](https://devexpress.github.io/testcafe/documentation/guides/advanced-guides/screenshots-and-videos.html) for the full list of available options).` work?

Reviewer: ```suggestion
TestCafe can capture videos and screenshots so you can debug your tests and examine page UI. You can record all tests or only those that failed, generate unique video files for each test or record all of them in one take, and single out specific interactions. (For a full list of available options, see the following help topic: [Screenshots and Videos](https://devexpress.github.io/testcafe/documentation/guides/advanced-guides/screenshots-and-videos.html)).
```

> TestCafe offers a number of debugging features, including the capability to automatically capture videos and screenshots during the testing process. You can record all your tests or only those that failed, generate multiple video files or stitch them together, and even single out specific interactions (see [the documentation](https://devexpress.github.io/testcafe/documentation/guides/advanced-guides/screenshots-and-videos.html) for the full list of available options).

Author: Suggestion: `TestCafe can capture videos and screenshots so you can debug your tests and examine the state of your pages' UI. `
Rationale: Clearer wording (there isn't necessarily something wrong with the page), getting rid of emotive language

> TestCafe offers a number of debugging features, including the capability to automatically capture videos and screenshots during the testing process. You can record all your tests or only those that failed, generate multiple video files or stitch them together, and even single out specific interactions (see [the documentation](https://devexpress.github.io/testcafe/documentation/guides/advanced-guides/screenshots-and-videos.html) for the full list of available options).

Reviewer 3: That's good - I'd probably even say "and examine page UI". Plural possessives are always awkward.

> But automating this process was not always easy. For a long time, Jenkins users had to manually match completed tests to their artifacts. The recent introduction of the [TestCafe Jenkins plugin](https://plugins.jenkins.io/testcafe/) changes that. Links to screenshots and videos taken during the test now automatically appear on the Jenkins test results page. You don't even need to keep your testing server running — all the necessary files are stored right inside the Jenkins build folder.

Reviewer 3: The first sentence is too "out of the blue", unexpected - we haven't said anything about automation yet... And now we go right in and say that it wasn't always "easy". Starting the entire thing with "but" is also not the best choice. 

```suggestion
You can easily review test screenshots and videos manually, but you'll mostly want recordings available somewhere within your automated testing process. Until recently, Jenkins users had to manually match completed tests to their artifacts. The newly implemented [TestCafe Jenkins plugin](https://plugins.jenkins.io/testcafe/) changes that. Links to screenshots and videos taken during a test run now automatically appear on the Jenkins test results page. You don't even need to keep your test server running — all the necessary files are stored right inside the Jenkins build folder.
```

Author: Suggestion: 
``` markdown
Many TestCafe users take advantage of Jenkins — a CI/CD solution — to automate their tests. Until recently, they had to manually match the screenshots and videos taken by TestCafe to the individual test reports. The newly implemented [TestCafe Jenkins plugin](https://plugins.jenkins.io/testcafe/) simplifies this process. 
```

Rationale: Sub-optimal logic: we start the paragraph by mentioning the process of manually reviewing test artifacts, then talk about automation, then go back to the manual review process.

Reviewer 3: That's good.

> But automating this process was not always easy. For a long time, Jenkins users had to manually match completed tests to their artifacts. The recent introduction of the [TestCafe Jenkins plugin](https://plugins.jenkins.io/testcafe/) changes that. The screenshots and videos taken during the test can now be automatically linked to on the Jenkins test results page. You don't even need to keep your testing server running — all the necessary files are stored right inside the Jenkins build folder.

Reviewer 4: > can now be automatically linked to

One general note I forgot yesterday - the correctors don't usually welcome putting verb + prep after the subject, like in this sentence, or `our test is going to run in.` below.

> Many TestCafe users take advantage of Jenkins — a CI/CD solution — to automate their tests. Until recently, they had to manually match the screenshots and videos taken by TestCafe to the individual test reports. The newly implemented [TestCafe Jenkins plugin](https://plugins.jenkins.io/testcafe/) simplifies this process. Links to screenshots and videos taken during the test now automatically appear on the Jenkins test results page. You don't even need to keep your testing server running — all the necessary files are stored right inside the Jenkins build folder.

Reviewer: ```suggestion
Many TestCafe users take advantage of Jenkins — a CI/CD solution — to automate their tests. Until recently, they had to manually match the screenshots and videos taken by TestCafe to the individual test reports. The newly implemented [TestCafe Jenkins plugin](https://plugins.jenkins.io/testcafe/) simplifies this process. Links to screenshots and videos taken during the test now automatically appear on the Jenkins test results page. There is no need to keep the testing server running — all required files are stored inside the Jenkins build folder.
```

<!--more-->

> **It looks something like this:**

Reviewer 4: Do we really need this emphasis (I mean bold font)? 

I would also go with a bit less colloquial phrase here:
`Your test results page will look like this:`

There are two reasons: 'it' is too vague here (what does it refer to?), and 'something' is too uncertain for a corporate blog (we should know for sure what our plugin renders).

> ![img](/testcafe/images/jenkins-plugin/1.png)

Reviewer 4: Please specify meaningful alternate text and file names for images.

> The test results page should look like this with our plugin enabled:

Reviewer: ```suggestion
With our plugin enabled, the test results page should appear as follows:
```

> # Test case: the disappearing button

Reviewer 4: This header should be level 2 `##`, because level 1 is `# Introducing the TestCafe Jenkins plugin`.

I guess you need to add a hash to all the headers here.

> Let's say, the 'Add to Cart' button keeps disappearing from the mobile version of your website.

Reviewer 3: 'Keeps disappearing' kind of implies that we already struggled with it for a while. I think Simple Present is a bit more straightforward here. 

```suggestion
Let's say the 'Add to Cart' button disappears on the mobile version of your website.
```

> Let's say the 'Add to Cart' button disappears on the mobile version of your website.

Reviewer: ```suggestion
Let's assume that the 'Add to Cart' button disappears on the mobile version of your website.
```

![The 'Add To Cart' button is absent from the mobile version of the website](/testcafe/images/jenkins-plugin/button-example.png)

> There are several ways TestCafe can help you detect this issue. You can:

Reviewer 3: The ways listed below are not about "detecting the issue", they only relate to emulating a mobile version. So I'd be specific here - this is not how we understand what's wrong, we're just setting up a test: 

```suggestion
There are several ways you can emulate a mobile device:
```

Author: Suggestion: 
```markdown
TestCafe includes support for [alt: supports] several mobile device emulation methods. You can:
Resize the browser window during testing
Run the test through a 3rd party cloud testing service
Take advantage of Chromium's built-in emulation mode
```
Rationale: We aren't talking about mobile emulation in general, just the methods supported by TestCafe. Changed the options' order, too (now it makes more sense)

Reviewer 3: Ok - that's fine too.

> TestCafe supports several mobile device emulation methods. You can:

Reviewer: ```suggestion
TestCafe supports the following mobile device emulation methods:
```

> * [Resize the browser window](https://devexpress.github.io/testcafe/documentation/guides/basic-guides/interact-with-the-page.html#resize-window) during testing

Reviewer: ```suggestion
* [Resize the browser window](https://devexpress.github.io/testcafe/documentation/guides/basic-guides/interact-with-the-page.html#resize-window) during the test
```

> * Run the test through a [3rd party cloud testing service](https://devexpress.github.io/testcafe/documentation/guides/concepts/browsers.html#browsers-in-cloud-testing-services)

Reviewer: ```suggestion
* Run the test through a [third-party cloud testing service](https://devexpress.github.io/testcafe/documentation/guides/concepts/browsers.html#browsers-in-cloud-testing-services)
```

> * [Resize the browser window](https://devexpress.github.io/testcafe/documentation/guides/basic-guides/interact-with-the-page.html#resize-window) during the test, or

Reviewer 3: I don't see the value of ", or".
```suggestion
* [Resize the browser window](https://devexpress.github.io/testcafe/documentation/guides/basic-guides/interact-with-the-page.html#resize-window) during the test
```

> * Take advantage of [Chromium's built-in emulation mode](https://devexpress.github.io/testcafe/documentation/guides/concepts/browsers.html#use-chromium-device-emulation)

Reviewer: ```suggestion
* Use [Chromium's built-in emulation mode](https://devexpress.github.io/testcafe/documentation/guides/concepts/browsers.html#use-chromium-device-emulation)
```

> For the sake of this tutorial, let's pick the third option, as it offers greater stability and speed.

Reviewer 3: "For the sake of" should be followed by something else - a word that describes how this choice affects the result. "For the sake of simplicity", "for the sake of completeness". "For the sake of this tutorial" doesn't really mean anything.

You also seem to imply that this third option is generally better than the other two... so here's my version:

```suggestion
The third option produces most stable and quick tests. We describe this option going forward and recommend that you use it if possible. 
```

Author: Suggestion:
```markdown
The third option offers greater speed and stability. We describe it going forward and recommend that you use it if possible.
```
Rationale: Potentially misleading verb choice (emulation cannot produce tests)

First, let's write a simple test that checks if the 'Add to Cart' button exists:

```js
test('`Add To Cart` button should exist', async t => {
    await t
        .click(Selector('button').withText('Add To Cart'));
});
```

> Now let's start automating.

Reviewer 3: I'd be more specific here. What are we trying to automate? E.g. "Now lets make sure we can run this test on Jenkins and review the results..." I don't know - something like that.

Author: How does `Now let's configure Jenkins to launch this test every time we build our project.` sound?

## Basic setup

> ### Step 1. Take care of the prerequisites

Reviewer: ```suggestion
### Step 1. Install prerequisites
```

> TestCafe requires a working Node.js installation to run. If your testing environment does not contain node.js, [this Jenkins plugin](https://plugins.jenkins.io/nodejs/) can install it for you. Make sure you have Chromium ready to go, too: TestCafe cannot function without a browser.

Reviewer 3: ```suggestion
TestCafe requires a working Node.js installation to run. If your testing environment does not contain node.js, [this Jenkins plugin](https://plugins.jenkins.io/nodejs/) can install it for you. Make sure you have Chromium too: TestCafe cannot function without a browser.
```

> TestCafe requires a working Node.js installation to run. If your testing environment does not contain node.js, [this Jenkins plugin](https://plugins.jenkins.io/nodejs/) can install it for you. Make sure you have Chromium too: TestCafe cannot function without a browser.

Reviewer: ```suggestion
TestCafe requires a working Node.js installation to run. If your testing environment does not contain node.js, the [Jenkins plugin](https://plugins.jenkins.io/nodejs/) can install it for you. Also ensure that Chromium is installed: TestCafe cannot function without a browser.
```

Author: If we're going to use the definite article, don't we need to specify which Jenkins plugin we're referring to? Users might assume that 'the Jenkins plugin' refers to the TestCafe Jenkins plugin, and that's not the plugin they need to use to install Node.js.

How does `If your testing environment does not contain node.js, the [Node.js Jenkins plugin](https://plugins.jenkins.io/nodejs/) can install it for you.` sound? `the eponymous Jenkins plugin` is also a decent option, imo.

Reviewer: `If your testing environment does not contain node.js, the [Node.js Jenkins plugin](https://plugins.jenkins.io/nodejs/) can install it for you.`

"the Node.js Jenkins plugin" sounds good.
And thanks for the clarification. I didn't realize that there were multiple versions of the Jenkins plugin...

### Step 2. Install the TestCafe Jenkins plugin

> To install the TestCafe Jenkins plugin, press the "Manage Plugins" link on the Manage Jenkins page, select the "Available" tab, and enter "testcafe" into the search field. Check the box next to our plugin and press the 'Install without restart' button below it.

Reviewer: I noticed that you say "press" multiple times (instead of "click"). Is this because you are talking about a mobile device? (Just checking.)

Author: See https://github.com/DevExpress/testcafe/pull/5329#discussion_r458429445

![Jenkins plugin search UI](/testcafe/images/jenkins-plugin/plugin-search-ui.png)

> ### Step 3. Install the necessary Node packages

Reviewer: ```suggestion
### Step 3. Install the required Node packages
```

Install the main testcafe library, as well as the [testcafe-reporter-jenkins](https://www.npmjs.com/package/testcafe-reporter-jenkins) package. The latter is needed to generate Jenkins-friendly JUnit format reports.

> To do this, add the following *build step* to your Jenkins project:

Reviewer: Instead of "To do this", specify to the reader what you're doing. (For example, "To install the testcafe-reporter-jenkins package, add the..."

Author: Gotcha. Can we just get rid of the first part of this sentence? Or switch the paragraphs? Or both:

```
Add the following build step to your Jenkins project:

npm i testcafe testcafe-reporter-jenkins

This command will install the main testcafe library, as well as the testcafe-reporter-jenkins package. The latter is needed to generate Jenkins-friendly JUnit format reports.
```

Reviewer: That's a good idea. How about:

`Add the following build step to your Jenkins project:

npm i testcafe testcafe-reporter-jenkins

This command installs the main testcafe library, as well as the testcafe-reporter-jenkins package. The latter is needed to generate Jenkins-friendly JUnit format reports.`

```bash
npm i testcafe testcafe-reporter-jenkins
```

### Step 4. Configure the tests

> You can configure your tests via:

Reviewer: ```suggestion
You can configure your tests via the following:
```

* the [command line interface](https://devexpress.github.io/testcafe/documentation/reference/command-line-interface.html)
* the JavaScript/TypeScript API
* the .testcaferc.json [configuration file](https://devexpress.github.io/testcafe/documentation/reference/configuration-file.html)

> The last option provides an easy, declarative way to define your test settings. You can store this file in the same folder as your tests.  Below is the configuration we're going to use in this tutorial:

Reviewer: ```suggestion
The last option provides an easy, declarative way to define your test settings. You can store this file in the same folder as your tests. Below is the configuration used in this tutorial:
```

```js
{
    "browsers": [
        "chromium:headless:emulation:device=iPhone 4",
        "chromium:headless:emulation:device=iPad Mini",
        "chromium:headless"
    ],
    "src": "test/e2e/**/*",
    "screenshots": {
        "takeOnFails": true,
        "fullPage": true
    },
    "videoPath": "videos",
    "reporter": {
        "name": "jenkins",
        "output": "report.xml"
    },
    "appCommand": "node server.js"
}
```

> The *Browser* array contains the list of browsers TestCafe will use to run our test. In it, we specified the mobile devices we want Chromium to emulate.

Reviewer 3: ```suggestion
The *Browser* array contains the list of browsers TestCafe will use to run our test. The example above specifies the mobile devices we want Chromium to emulate.
```

Author: Suggestion:
``` markdown 
The Browser array contains the list of browsers TestCafe will use to run our test. The [[TestCafe browser syntax](link to docs)] helped us specify the mobile devices we want Chromium to emulate.
```

Rationale: The reader might want to know how exactly these devices were specified

The *name* property of the *reporter* object is set as *jenkins*, which ensures that the reports generated by TestCafe can be properly parsed by the server.

If you decide against using a config file, don't forget to [manually set](https://devexpress.github.io/testcafe/documentation/reference/command-line-interface.html#-r-nameoutput---reporter-nameoutput) the report format when launching your tests, like so:

```bash

> testcafe chrome test.js -r jenkins:report.xml.

Reviewer 2: Extra ```.``` at the end of the line

> If you decide against using a config file, don't forget to [manually set](https://devexpress.github.io/testcafe/documentation/reference/command-line-interface.html#-r-nameoutput---reporter-nameoutput) the report format when launching your tests, like so:

Reviewer: ```suggestion
If you decide against using a config file, be sure to [manually set](https://devexpress.github.io/testcafe/documentation/reference/command-line-interface.html#-r-nameoutput---reporter-nameoutput) the report format when launching your tests as follows:
```

```

> Otherwise, just run the following command:

Reviewer: ```suggestion
Otherwise, simply run the following command:
```

```bash
npx testcafe
```

### Step 5. Configure the reporter

> Add the "Publish JUnit test result report" *post-build action* to turn the pre-included [JUnit](https://plugins.jenkins.io/junit/) reporter plugin on. To display screenshots and videos alongside your test results, select the "Include links to TestCafe artifacts" option from the "Additional test report features" drop-down.

Reviewer 3: ```suggestion
Turn on the [JUnit](https://plugins.jenkins.io/junit/) reporter plugin: add the "Publish JUnit test result report" *post-build action*. To display screenshots and videos alongside your test results, select the option "Include links to TestCafe artifacts" from the "Additional test report features" drop-down.
```

> Turn on the [JUnit](https://plugins.jenkins.io/junit/) reporter plugin: add the "Publish JUnit test result report" *post-build action*. To display screenshots and videos alongside your test results, select the option "Include links to TestCafe artifacts" from the "Additional test report features" drop-down.

Reviewer: Do you mean:

`To turn on the [JUnit](https://plugins.jenkins.io/junit/) reporter plugin, add the "Publish JUnit test result report" *post-build action*. To display screenshots and videos alongside your test results, select the "Include links to TestCafe artifacts" option from the "Additional test report features" drop-down menu.
```

Author: Not quite!
I'm reluctant to start the first sentence with the word 'To' because the necessity of turning the JUnit reporter plugin on is not self-evident to the reader. The imperative mood makes it clear that this seemingly unrelated intermediate step is, in fact, mandatory.

The same logic does not apply to the second sentence because, while unlikely, it is possible that the user might want to leave the feature we're describing disabled. Besides, skipping this step will not break their Jenkins setup.

As far as the word order is concerned, @abadzhev changed the second sentence during [an earlier review](https://github.com/DevExpress/testcafe/pull/5329#discussion_r456609406), and I'll be happy to change it back :)

Reviewer: Thanks for the clarification. The second sentence is fine.
I've made a minor change in the suggestion below.

Reviewer: ```suggestion
Turn on the [JUnit](https://plugins.jenkins.io/junit/) reporter plugin: add the "Publish JUnit test result report" *post-build action*. To display screenshots and videos alongside your test results, select the "Include links to TestCafe artifacts" option from the "Additional test report features" drop-down menu.
```

> Important: Don't forget to check the "Retain long standard output/error" box. Otherwise the plugin will automatically truncate information concerning test screenshots and videos.

Reviewer: ```suggestion
Important: Remember to check the "Retain long standard output/error" box. Otherwise, the plugin will automatically truncate information about test screenshots and videos.
```

Author: I probably need to rewrite the second sentence. Thing is: Jenkins reports generated without this option enabled feature incomplete (and, therefore, useless) URLs. It's not the metadata that gets truncated, as the current phrasing (`information about test screenshots`) would have you believe, but the most important part of the actual data.

Would `Otherwise, the reporter plugin will automatically truncate the URLs of your videos and screenshots.` work?

Reviewer: `Otherwise, the reporter plugin will automatically truncate the URLs of your videos and screenshots.`

Sounds good. Please apply the change.

![JUnit reporter configuration screen](/testcafe/images/jenkins-plugin/junit-reporter-configuration.png)

### Step 6. Run the build

> Save the changes to your project and run the build by pressing the "Build now" button on the project page:

Reviewer 3: The title says "Run the build"... The command is "Build now". This may be a distinction without a difference in that Jenkins thing... But those are usually two different things. Just want to make sure we say the right things. 

```suggestion
Save the changes to your project and press the "Build now" button on the project page:
```

![Jenkins Project menu with the Build Now Button highlighted](/testcafe/images/jenkins-plugin/build-now-button.png)

> ## Looking at the results

Reviewer 3: ```suggestion
## Review the results
```

> Save the changes to your project and press the "Build now" button on the project page:

Reviewer: Do you say "press" because you are talking about a mobile device?

Author: Thanks for pointing this out! I was under the impression that the verb _to press_ is used to describe interactions with UI elements that enter a visually distinct 'pressed' state when clicked. Upon consulting two different style guides, I realized that I was mistaken, and that I should only really use the word 'press' when talking about physical I/O devices (i.e. the keyboard or the touch screen). Is that correct?

Reviewer: Correct, press should only be used when you are actually "pressing" something, like the Shift key on a keyboard or a button on a tablet.

I think "press" is used in other parts of the article as well. Can you replace them?

> After the tests are completed, press this link on the build page to view the results:

Reviewer 3: ```suggestion
After the tests are completed, press the following link on the build page to view the results:
```

![Test results link](/testcafe/images/jenkins-plugin/test-results-link.png)

> There you'll see links to screenshots and videos taken during the test.

Reviewer: ```suggestion
You will see links to screenshots and videos taken during the test.
```

![Test results view](/testcafe/images/jenkins-plugin/test-results-view.png)

> If you were working on fixing the cart button, these artifacts would instantly indicate if it's there or not.

Reviewer 3: ```suggestion
These artifacts would instantly indicate if the Cart button was present on the page.
```

Author: Suggestion: `These artifacts instantly indicate if the Cart button is present on the page.`
Rationale: Makes sense to use the zero conditional when describing a general rule

Reviewer 3: Sounds good.

## Jenkins Pipeline integration

> If you don't want to use the Jenkins GUI to launch these tests, you can always make use of the [Jenkins Pipeline](https://www.jenkins.io/doc/book/pipeline/). It lets you declaratively configure your CI/CD setup via Jenkinsfile, a text file checked into the project's source control repository. Here's a sample Jenkinsfile to be used with TestCafe:

Reviewer 3: If "Jenkins Pipeline" is a name, doesn't require an article. Jenkinsfile seems to be referred as an instance, so I'd use an indefinite article (you do that in the last sentence anyway).

```suggestion
If you don't want to use the Jenkins GUI to launch these tests, you can always make use of [Jenkins Pipeline](https://www.jenkins.io/doc/book/pipeline/). It lets you declaratively configure your CI/CD setup via a Jenkinsfile, a text file checked into the project's source control repository. Here's a sample Jenkinsfile to be used with TestCafe:
```

> If you don't want to use the Jenkins GUI to launch these tests, you can always make use of the [Jenkins Pipeline](https://www.jenkins.io/doc/book/pipeline/). It lets you declaratively configure your continuous delivery setup via Jenkinsfile, a text file checked into the project's source control repository. Here's a sample Jenkinsfile to be used with TestCafe:

Reviewer 2: ```Jenkinsfile``` can be used not only for ```continuous delivery```, but also for ```continuous integration``` setup (```CI/CD Pipeline```).
I feel that mentioning only ```continuous delivery``` may confuse a reader.

> If you don't want to use the Jenkins GUI to launch these tests, you can always make use of [Jenkins Pipeline](https://www.jenkins.io/doc/book/pipeline/). It lets you declaratively configure your CI/CD setup via a Jenkinsfile, a text file checked into the project's source control repository. Here's a sample Jenkinsfile to be used with TestCafe:

Reviewer: Can you clarify what you mean by "checked into"? Is "checked" the correct technical term in this context?

Author: 'Checking a file into source control' means instructing a source control management system of your choice to track this file going forward.

https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes : `<...> tell Git how to diff non-text files, or have Git filter content before you check it into or out of Git.`

https://devcenter.heroku.com/articles/git: `Common causes of large repositories are binary files checked into the repository  (Git is notoriously bad at handling binaries) or constantly-changing development logs.`

Reviewer: Thanks for the clarification. Please see my new suggestion below.

Reviewer: ```suggestion
If you do not wish to use the Jenkins GUI to launch these tests, you can always use the [Jenkins Pipeline](https://www.jenkins.io/doc/book/pipeline/). It lets you declaratively configure your CI/CD setup via a Jenkinsfile - a text file checked into the project's source control repository. The following is a sample Jenkinsfile to be used with TestCafe:
```

```java
pipeline {
    agent any

    stages {
        stage('Run E2E tests') {
            steps {
                sh 'npm install'
                sh 'npm run test-e2e'
            }
            post {
                always {
                          junit keepLongStdio: true,
                          testDataPublishers: [[$class: 'TestCafePublisher']],
                          testResults: '*.xml'
                }
            }
        }
    }
}
```

> Note: The `test-e2e` npm script in this example is responsible for launching TestCafe. See the [TestCafe CLI](https://devexpress.github.io/testcafe/documentation/reference/command-line-interface.html) documentation for more information on the subject, and don't forget to manually set the reporter option as outlined earlier in this article.

Reviewer: ```suggestion
Note: The `test-e2e` npm script in this example launches TestCafe. As mentioned earlier in this topic, be sure to manually set the reporter option. For more information on the TestCafe CLI, see the following help topic: [Command Line Interface](https://devexpress.github.io/testcafe/documentation/reference/command-line-interface.html).
```

--

> If you enjoyed this guide, you might like our other articles on the subject of debugging: [How to Debug Tests in TestCafe: Quick Guide](https://devexpress.github.io/testcafe/media/team-blog/how-to-debug-tests-in-testcafe-quick-guide.html) and [How to speed up debugging in TestCafe: Tips and Tricks](https://devexpress.github.io/testcafe/media/team-blog/how-to-speed-up-debugging-in-testcafe-tips-and-tricks.html).

Reviewer: ```suggestion
If you enjoyed this guide, you may like our other articles on debugging: [How to Debug Tests in TestCafe: Quick Guide](https://devexpress.github.io/testcafe/media/team-blog/how-to-debug-tests-in-testcafe-quick-guide.html) and [How to speed up debugging in TestCafe: Tips and Tricks](https://devexpress.github.io/testcafe/media/team-blog/how-to-speed-up-debugging-in-testcafe-tips-and-tricks.html).
```

Author: Can I substitute `help topics` for `guides`? These aren't really help topics in the proper sense, as they aren't a part of a knowledge base. This is what I imagine when I hear the words 'help topic":
![image](https://user-images.githubusercontent.com/43554315/88168148-abb5e380-cc22-11ea-9d87-463fef43a262.png)

Reviewer: As far as I know, "help topic" is the phrase that Ray prefers, so I think we should follow his guidelines.
